### ISSUE TITLE
Nvidia Nvenc hardware accelleration on docker container give PlaybackErrorNoCompatibleStream

### ISSUE THREAD

**Issue by antipiot** [jellyfin/jellyfin#2395]
Hello!

I'm trying to achieve nvenc harware acceleration on my jellyfin installation:
At the moment: driver is installed and looks to be working fine on host:

<details><summary>nvidia-smi</summary>
`"root@openmediavault:~# nvidia-smi
Mon Feb 10 15:55:38 2020
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 418.74       Driver Version: 418.74       CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 760     Off  | 00000000:01:00.0 N/A |                  N/A |
|  0%   41C    P0    N/A /  N/A |      0MiB /  1999MiB |     N/A      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0                    Not Supported                                       |
+-----------------------------------------------------------------------------+"
</details>

It  looks to be working aswell trough docker container - all dependencies met.

<details><summary>docker nvidia/cuda nvidia-smi</summary>

`docker run --gpus all nvidia/cuda nvidia-smi
Mon Feb 10 14:56:50 2020
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 418.74       Driver Version: 418.74       CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 760     Off  | 00000000:01:00.0 N/A |                  N/A |
|  0%   42C    P0    N/A /  N/A |      0MiB /  1999MiB |     N/A      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0                    Not Supported                                       |
+-----------------------------------------------------------------------------+`
</details>

Jellyfin start command looks like:

<details><summary>docker run</summary>
`	docker run -d \
  --privileged \
  --name=jellyfin \
  -e PUID=1000 \
  -e PGID=100 \
  -e "NVIDIA_DRIVER_CAPABILITIES=all" \
  -e NVIDIA_VISIBLE_DEVICES=all \
  --gpus all \
  --runtime=nvidia \
  -p 8096:8096 \
  -p 8920:8920 \
  -v /sharedfolders/data/docker/jellyfin/config:/config \
  -v /sharedfolders/data/docker/jellyfin/movies:/media \
  -v /sharedfolders/data/docker/jellyfin/cache:/cache \
  --restart unless-stopped \
	jellyfin/jellyfin`

</details>
but, once i try to start a playback who need to be transcoded i get this message:
`PlaybackErrorNoCompatibleStream`

and i can see this at the bottom of the ffmpeg transcode log:
`[hevc_cuvid @ 0x56532b687440] Cannot load libnvcuvid.so.1
[hevc_cuvid @ 0x56532b687440] Failed loading nvcuvid.
Stream mapping:
  Stream #0:0 -> #0:0 (hevc (hevc_cuvid) -> h264 (h264_nvenc))
  Stream #0:1 -> #0:1 (truehd (native) -> aac (native))
Error while opening decoder for input stream #0:0 : Operation not permitted`

i've already read everything i could find on this sugnet but could not get it working.

Do you know what's going on? what  did i missed?

Many thanks for your time and work!

Regards

Jonas

**Comment by Shadowghost** [comment]
Short explanation why it does not work:
ffmpeg tries to decode HEVC with your nvidia graphics card which does not support it (take a look [here](https://developer.nvidia.com/video-encode-decode-gpu-support-matrix)).

#2127 aims to fix something similar for >8bit video instead of not supported stream codec.
I'm not sure how this is currently handled by ffmpeg or the jellyfin logic, but there should be a fallback to software decoding when the original codec can't be decoded in hardware.

Could also be that it's already fixed on current nightly builds, there were a lot of changes to hw en/decoding since 10.4.3. Maybe you can give them a shot.

**Comment by antipiot** [comment]
Many thanks for your quick and accurate answer.

Tried with the latest nightly  with no luck:

 _STATISTICS_TAGS-eng: BPS DURATION NUMBER_OF_FRAMES NUMBER_OF_BYTES
[hevc_cuvid @ 0x558686136680] Cannot load libnvcuvid.so.1
[hevc_cuvid @ 0x558686136680] Failed loading nvcuvid.
Stream mapping:
  Stream #0:0 -> #0:0 (hevc (hevc_cuvid) -> h264 (h264_nvenc))
  Stream #0:1 -> #0:1 (truehd (native) -> aac (native))

Guess i have else to buy a GPU supporting HEVC or to wait:-)

Anyway - thanks again!

**Comment by Shadowghost** [comment]
Well that's another issue you got there. Seems like some nvidia drivers/libs are missing on the image.

**Comment by Shadowghost** [comment]
Check you docker setup and try again with h264 source material only (see comment above).
Reading the log you provided it seems like it still tries to decode HEVC with your graphics card (IMHO you should file another bugreport for that).

**Comment by antipiot** [comment]
Hey!
Finally got it working fine with x264 files: GPU is now correctly encoding and decoding files:

On the host side, i had to install libnvcuvid1 wich was missing:
apt-get install -t stretch-backports libnvcuvid1

then you can check both host and container have correct access to it:

```
ldconfig -p | grep cuvid
docker exec -it jellyfin-nightly ldconfig -p | grep cuvid
```

both should return value as:
`libnvcuvid.so.1 (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libnvcuvid.so.1`

then got a new error:
`h264_nvenc @ 0x55855d101140] Cannot load libnvidia-encode.so.1
[h264_nvenc @ 0x55855d101140] The minimum required Nvidia driver for nvenc is 418.30 or newer`

Checked driver wich was ok so i went on downloading missing lib:
`apt-get install -t stretch-backports libnvidia-encode1`

Again, you can check container has access with:
ldconfig -p | grep  libnvidia-encode.so.1
docker exec -it jellyfin-nightly ldconfig -p | grep  libnvidia-encode.so.1

TL:DR:
Download missing libs: 
`apt-get install -t stretch-backports libnvidia-encode1
apt-get install -t stretch-backports libnvcuvid1`
